tpm.go

```diff
@@ -0,0 +1,937 @@
+// of the key associated with the verified endorsement key certificate. The session key can only be retrieved by and used on the TPM
+// for which the endorsement certificate was issued. If the connection was created with ConnectToDefaultTPM, the session may be
+// salted with a value protected by the public part of the endorsement key if one exists or one is able to be created, but as the key
+// is not associated with a verified credential, there is no guarantee that only the TPM is able to retrieve the session key.
+func (t *TPMConnection) HmacSession() tpm2.SessionContext {
+	if t.hmacSession == nil {
+		return nil
+	}
+	return t.hmacSession.WithAttrs(tpm2.AttrContinueSession)
+}
+
+func (t *TPMConnection) Close() error {
+	t.FlushContext(t.hmacSession)
+	return t.TPMContext.Close()
+}
+
+// createTransientEk creates a new primary key in the endorsement hierarchy using the default RSA2048 EK template.
+func createTransientEk(tpm *tpm2.TPMContext) (tpm2.ResourceContext, error) {
+	session, err := tpm.StartAuthSession(nil, tpm.EndorsementHandleContext(), tpm2.SessionTypeHMAC, nil, tpm2.HashAlgorithmSHA256)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot start auth session: %w", err)
+	}
+	defer tpm.FlushContext(session)
+
+	ek, _, _, _, _, err := tpm.CreatePrimary(tpm.EndorsementHandleContext(), nil, &ekTemplate, nil, nil, session)
+	return ek, err
+}
+
+// verifyEk verifies that the public area of the ResourceContext that was read back from the TPM is associated with the supplied EK
+// certificate. It does this by obtaining the public key from the EK certificate, inserting it in to the standard EK template,
+// computing the expected name of the EK object and then verifying that this name matches the result of ResourceContext.Name. This
+// works because go-tpm2 cross-checks that the name and public area returned from TPM2_ReadPublic match when initializing the
+// ResourceContext.
+//
+// Success confirms that the ResourceContext references the public area associated with the public key of the supplied EK certificate.
+// If that certificate has been verified, the ResourceContext can safely be used to encrypt secrets that can only be decrpyted and
+// used by the TPM for which the EK certificate was issued, eg, for salting an authorization session that is then used for parameter
+// encryption.
+func verifyEk(cert *x509.Certificate, ek tpm2.ResourceContext) error {
+	// Obtain the RSA public key from the endorsement certificate
+	pubKey, ok := cert.PublicKey.(*rsa.PublicKey)
+	if !ok {
+		return errors.New("cannot obtain RSA public key from certificate")
+	}
+
+	// Insert the RSA public key in to the EK template to compute the name of the EK object we expected to read back from the TPM.
+	var ekPublic *tpm2.Public
+	b, _ := tpm2.MarshalToBytes(ekTemplate)
+	tpm2.UnmarshalFromBytes(b, &ekPublic)
+
+	if pubKey.E != 65537 {
```

Comment:
Maybe a comment here similar to the one below in https://github.com/snapcore/secboot/pull/1/files#diff-7d1c2a3334601b6c1958aae0a594cba8R72

Reply:
I've added a brief comment here now.